session/tokenizer.go

```diff
@@ -0,0 +1,147 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package session
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/gofrs/uuid"
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/pkg/errors"
+	"github.com/tidwall/gjson"
+
+	"github.com/ory/herodot"
+	"github.com/ory/kratos/driver/config"
+	"github.com/ory/kratos/x"
+	"github.com/ory/x/fetcher"
+	"github.com/ory/x/jsonnetsecure"
+	"github.com/ory/x/jwksx"
+	"github.com/ory/x/otelx"
+)
+
+type (
+	tokenizerDependencies interface {
+		jsonnetsecure.VMProvider
+		x.TracingProvider
+		x.HTTPClientProvider
+		config.Provider
+		x.JWKFetchProvider
+	}
+	Tokenizer struct {
+		r       tokenizerDependencies
+		nowFunc func() time.Time
+	}
+	TokenizerProvider interface {
+		SessionTokenizer() *Tokenizer
+	}
+)
+
+func NewTokenizer(r tokenizerDependencies) *Tokenizer {
+	return &Tokenizer{r: r, nowFunc: time.Now}
+}
+
+func (s *Tokenizer) SetNowFunc(t func() time.Time) {
+	s.nowFunc = t
+}
+
+func (s *Tokenizer) TokenizeSession(ctx context.Context, template string, session *Session) (err error) {
+	ctx, span := s.r.Tracer(ctx).Tracer().Start(ctx, "sessions.ManagerHTTP.TokenizeSession")
+	defer otelx.End(span, &err)
+
+	tpl, err := s.r.Config().TokenizeTemplate(ctx, template)
+	if err != nil {
+		return err
+	}
+
+	httpClient := s.r.HTTPClient(ctx)
+	key, err := s.r.Fetcher().ResolveKey(
+		ctx,
+		tpl.JWKSURL,
+		jwksx.WithCacheEnabled(),
+		jwksx.WithCacheTTL(time.Hour),
+		jwksx.WithHTTPClient(httpClient))
+	if err != nil {
+		if errors.Is(err, jwksx.ErrUnableToFindKeyID) {
+			return errors.WithStack(herodot.ErrBadRequest.WithReasonf("Could not find key a suitable key for tokenization in the JWKS url."))
+		}
+		return err
+	}
+
+	alg := jwt.GetSigningMethod(key.Algorithm())
+	if alg == nil {
+		return errors.WithStack(herodot.ErrBadRequest.WithReasonf("The JSON Web Key must include a valid \"alg\" parameter but \"%s\" was given.", key.Algorithm()))
+	}
+
+	vm, err := s.r.JsonnetVM(ctx)
+	if err != nil {
+		return err
+	}
+
+	fetch := fetcher.NewFetcher(fetcher.WithClient(httpClient))
+
+	now := s.nowFunc()
+	token := jwt.New(alg)
+	token.Header["kid"] = key.KeyID()
+	claims := jwt.MapClaims{
+		"jti": uuid.Must(uuid.NewV4()).String(),
+		"iss": s.r.Config().SelfPublicURL(ctx).String(),
+		"exp": now.Add(tpl.TTL).Unix(),
+		"sub": session.IdentityID.String(),
+		"sid": session.ID.String(),
+		"nbf": now.Unix(),
+		"iat": now.Unix(),
+	}
+
+	if mapper := tpl.ClaimsMapperURL; len(mapper) > 0 {
+		jn, err := fetch.FetchContext(ctx, mapper)
+		if err != nil {
+			return err
```

Comment:
no `errors.WithStack`?

Reply:
`fetch.FetchContext` already adds stack context afaik since it's in ory/x